### DISCUSSION TITLE
Custom CosmosSerializer and use GetItemLinqQueryable

### DISCUSSION THREAD

**Discussion by mateuszkumpf** [IEvangelist/azure-cosmos-dotnet-repository#321] [Q&A]
Hey guys,

I created my custom CosmosSerializer:

```csharp
public class DefaultCosmosSerializer : CosmosSerializer
{
    private readonly JsonSerializer _jsonSerializer;

    public DefaultCosmosSerializer()
    {
        var settings = new JsonSerializerSettings();

        SerializerSettingsHelper.ConfigureSerializerSettingsForDatabase(settings);

        _jsonSerializer = JsonSerializer.CreateDefault(settings);
    }

    public override T FromStream<T>(Stream stream)
    {
        using (var streamReader = new StreamReader(stream))
        {
            return _jsonSerializer.Deserialize<T>(new JsonTextReader(streamReader))!;
        }
    }

    public override Stream ToStream<T>(T input)
    {
        using (var stringWriter = new StringWriter())
        {
            using (var jsonTextWriter = new JsonTextWriter(stringWriter))
            {
                _jsonSerializer.Serialize(jsonTextWriter, input, typeof(T));
                var result = stringWriter.ToString();

                return new MemoryStream(Encoding.UTF8.GetBytes(result));
            }
        }
    }
}
```

For JsonSerializerSettings I set ContactResolver as CamelCasePropertyNamesContractResolver().
In the end I override default serializer using AdditionalSetupAction in AddCosmosRepository() like that

```csharp
private static Action<CosmosClientOptions> AdditionalSetupAction(CosmosDbAccountSettings cosmosDbAccountSettings)
{
    return opt =>
    {
        opt.SerializerOptions = null;
        opt.Serializer = new DefaultCosmosSerializer();
        opt.MaxRetryAttemptsOnRateLimitedRequests = cosmosDbAccountSettings.MaxRetryAttemptsOnRateLimitedRequests;
    };
}
```

What is my problem? I can't use every place where the repository uses GetItemLinqQueryable() (e.g. GetAsync() with expression), because it builds the database query using the field names in the class as they are stored (in the PascalCase policy). The only fields I can use are those that have a JsonProperty attribute and a filled PropertyName or there camelCase policy selected.
Since I used my own serialiser (I needed more options which CosmosSerialisationOptions does not contain) I am forced to create SQL queries myself and use GetByQueryAsync() which, with modifications to the classes, is quite difficult to maintain.

I would ask for some advice.

Regards,
Mateusz Kumpf

**Comment by IEvangelist** [comment]
Yuck, that's unfortunate that it's not honoring your custom JSON serializer. I'll dig into this as soon as I am able and report back.

**Reply by mateuszkumpf** [reply]
It looks to me like Microsoft in their implementation of GetItemLinqQueryable() honoring only SerilizeOptions, if this object is null (has to be null if I want to use custom CosmosSerializer) then they choose default policy (so PascalCase).
I'm not sure it is correct way but you could set linqSerializerOptions based on RepositorySerializationOptions.PropertyNamingPolicy but then everyone which uses this repository should remember that in addition to setting your own serializer, also set the same policy on the RepositoryOptions.RepositorySerializationOptions object.
